### PR TITLE
viomi: ViomiMapSegmentEditCapability

### PR DESCRIPTION
**Pull request by jloewe**
## Type of change

- [x] Capability implementation for existing core capability

# Description

I implemented ViomiMapSegmentEditCapability for viomi vacuums
The implementation was tested on my viomi.v7 and works great.

### CODE REVIEW

File: lib/robots/viomi/capabilities/ViomiMapSegmentEditCapability.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+    /**
+     * @param {import("../../../entities/core/ValetudoMapSegment")} segmentA
+     * @param {import("../../../entities/core/ValetudoMapSegment")} segmentB
+     * @returns {Promise<void>}
+     */
+    async joinSegments(segmentA, segmentB) {
+        await this.robot.sendCommand("arrange_room", {
+            lang: this.lang,
+            mapId: this.robot.currentMapId,
```
Comment (Hypfer) [inline lib/robots/viomi/capabilities/ViomiMapSegmentEditCapability.js:45]: Will this ever be anything that isn't `1`?

Comment (jloewe) [reply lib/robots/viomi/capabilities/ViomiMapSegmentEditCapability.js:45]: The currentMapId? For me it currently has a value that looks something like this: `1610520334`

Comment (Hypfer) [reply lib/robots/viomi/capabilities/ViomiMapSegmentEditCapability.js:45]: Oh that's the unix timestamp for `Wed Jan 13 2021 06:45:34 GMT+0000`

That correlates with the first time you've used the robot I'd assume?

Comment (jloewe) [reply lib/robots/viomi/capabilities/ViomiMapSegmentEditCapability.js:45]: Yep that should match up

Comment (Hypfer) [reply lib/robots/viomi/capabilities/ViomiMapSegmentEditCapability.js:45]: Does it fail if it's not set?

Comment (jloewe) [reply lib/robots/viomi/capabilities/ViomiMapSegmentEditCapability.js:45]: Yes, I just tried it and the robot really seems not to like it. The response also takes forever.

File: lib/robots/viomi/ViomiValetudoRobot.js [hunk 2]
```diff
@@ -381,6 +391,10 @@ class ViomiValetudoRobot extends MiioValetudoRobot {
             }));
         }
 
+        if (data["cur_mapid"] !== undefined) {
```
Comment (Hypfer) [inline lib/robots/viomi/ViomiValetudoRobot.js:394]: I'm not a fan of this.
With this, we're at two dynamic viomi properties :/

File: lib/robots/viomi/capabilities/ViomiMapSegmentEditCapability.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        return `${roundedX}_${roundedY}`;
+    }
+
+    /**
+     * @param {import("../../../entities/core/ValetudoMapSegment")} segmentA
+     * @param {import("../../../entities/core/ValetudoMapSegment")} segmentB
+     * @returns {Promise<void>}
+     */
+    async joinSegments(segmentA, segmentB) {
+        await this.robot.sendCommand("arrange_room", {
```
Comment (Hypfer) [inline lib/robots/viomi/capabilities/ViomiMapSegmentEditCapability.js:43]: If the currentMapId is required for this to work, it should throw if it's not available.

Another approach could be to poll the status before. Not sure what would be the best approach here though

File: lib/robots/viomi/capabilities/ViomiMapSegmentEditCapability.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+     * @param {object} pA
+     * @param {number} pA.x
+     * @param {number} pA.y
+     * @param {object} pB
+     * @param {number} pB.x
+     * @param {number} pB.y
+     * @returns {Promise<void>}
+     */
+    async splitSegment(segment, pA, pB) {
+        await this.robot.sendCommand("arrange_room", {
```
Comment (Hypfer) [inline lib/robots/viomi/capabilities/ViomiMapSegmentEditCapability.js:66]: Same thing applies to split as well

**Review by Hypfer — CHANGES_REQUESTED** [review]

File: lib/robots/viomi/ViomiValetudoRobot.js (hunk 2 above)
Comment (jloewe) [reply lib/robots/viomi/ViomiValetudoRobot.js:394]: For some reason `get_props` with the `cur_mapid` is also slow on my robot. Is the mapId contained in the uploaded map that is parsed by the ViomiMapParser?

Comment (jloewe) [reply lib/robots/viomi/ViomiValetudoRobot.js:398]: I think I found it in my map dump. The mapId starts at offset 0x04 and is four byte long.

Comment (Hypfer) [reply lib/robots/viomi/ViomiValetudoRobot.js:398]: In that case, you can add that to the metaData field of the ValetudoMap returned by the map parser and access that via the robot

plus this check to fail when the map is missing

`if (this.robot.state.map?.metaData?.defaultMap !== true) {`

Comment (jloewe) [reply lib/robots/viomi/ViomiValetudoRobot.js]: Done. I had to [modify the ValetudoMap to take the metaData option](https://github.com/Hypfer/Valetudo/pull/766/files#diff-40a906408dc9f223acf527c14e56488a5a1e4d7a26fff75f0585ff97f4ed8a3eL34).

File: lib/entities/map/ValetudoMap.js [hunk 5]
```diff
@@ -31,7 +31,10 @@ class ValetudoMap extends SerializableEntity { //TODO: Current, Historic, Etc.
         this.layers = [];
         this.entities = [];
 
-        this.metaData.version = 1; //Will probably be incremented some day
+        this.metaData = {
```
Comment (Hypfer) [inline lib/entities/map/ValetudoMap.js:34]: Are you sure that this is required?

Comment (jloewe) [reply lib/entities/map/ValetudoMap.js:34]: I will check it again, but I'm pretty sure

Comment (jloewe) [reply lib/entities/map/ValetudoMap.js]: Oh sorry I didn't realize that the SerializableEntity handles the metaData

**Pull request merged**